compiler/src/main/java/com/squareup/anvil/compiler/codegen/PsiUtils.kt

```diff
@@ -159,3 +153,94 @@ private fun KtClassOrObject.findScopeClassLiteralExpression(
           element = this
       )
 }
+
+internal fun PsiElement.fqNameOrNull(
+  module: ModuleDescriptor
+): FqName? {
+  // Usually it's the opposite way, the require*() method calls the nullable method. But in this
+  // case we'd like to preserve the better error messages in case something goes wrong.
+  return try {
+    requireFqName(module)
+  } catch (e: AnvilCompilationException) {
+    null
+  }
+}
+
+internal fun PsiElement.requireFqName(
+  module: ModuleDescriptor
+): FqName {
+  val containingKtFile = parentsWithSelf
+      .filterIsInstance<KtClassOrObject>()
+      .first()
+      .containingKtFile
+
+  fun failTypeHandling(): Nothing = throw AnvilCompilationException(
+      "Don't know how to handle Psi element: $text",
+      element = this
+  )
+
+  val classReference = when (this) {
+    // If a fully qualified name is used, then we're done and don't need to do anything further.
+    is KtDotQualifiedExpression -> return FqName(text)
+    is KtNameReferenceExpression -> getReferencedName()
+    is KtUserType -> referencedName ?: failTypeHandling()
+    is KtTypeReference -> text
+    is KtNullableType -> return innerType?.requireFqName(module) ?: failTypeHandling()
```

Comment:
Should we be using `fqNameOrNull()` in case if type is present to get better error message from `failTypeHandling`?

Reply:
That's not necessary here. If `innerType` is present, then it must have a fqName.